### ISSUE TITLE
Problem when executing eagerly

### ISSUE THREAD

**Issue by nguerinjr** [tensorflow/compression#9]
Hello!

I'm trying to use some of your layers, but I see they're not fully compatible with eager execution (even though I see some if's to treat cases of eager execution in your code).

One example is the `self.add_loss` in the entropy bottleneck layer.

I've started reading the keras code, and I've seen this commentary:

> Note that `add_loss` is not supported when executing eagerly. Instead,
>     variable regularizers may be added through `add_variable`. Activity
>     regularization is not supported directly (but such losses may be returned
>     from `Layer.call()`).

I don't know if simply using the `add_variable` instead would fix this problem.

I have not tried any solution yet. I've just wanted to share the problem with the code maintainer. 

Maybe anybody seeing this can point out a solution before I do.

; )

**Comment by jonycgn** [comment]
Hi, thanks for reporting this.

Correct, we haven't had a use case for eager mode, so we don't currently support it yet. There's some if branches that were borrowed from other code, but no one has tested that.

I'm not sure the loss that's added there could be set up correspondingly by passing a regularizer object when the layer's variables are created. It's not a simple loss applied to each variable, but combines several variables, as well as the training data. It's is used for estimating the quantiles of the density model, which are necessary to create the probability tables for the range coder.

What's your use case for eager mode in this case?

**Comment by nguerinjr** [comment]
Hello @jonycgn !

Sorry for the lack of an answer. Actually I haven't perceived the message later. I've started studying other models and didn't go back to your framework.

Recently, I've started looking again at your work, looked at your repo, and saw the message. As I see now, one of the problems in eager exec is the None gradients for your reparametrization variables in GDN, for example (both in 1.x and 2.x). 

I'm trying to use some of your layers with TF 1.15. I intend to upgrade the code to 2.0 in the future (don't know when). So if I can, I may submit a pull request to you (if you do not do it before me and if I get minimal changes for it). Do you intend to upgrade the code to the 2.0 version?

Currently, I'm building a .so of my own since the one in your pip package is not working for me (the pmf initializer in the entropy bottleneck returns a SIGSEGV). A compatibility issue, maybe. 

I'm still trying to understand why, when encapsulating all your layers code in a Keras model, the entropy bottleneck matrixes returns None as the gradient. The parameters of that layer are trainable, I presume. Also, when I try to execute one of the implementations of your work, with the .so I've built locally, I have a weird failure when loading the cuDNN algorithm to perform convolutions. Maybe other compatibility issues.

Maybe, after I understand the behavior of the gradients I can think of a change to work with eager execution.

**Comment by jonycgn** [comment]
Hi,

thanks for your feedback. I just confirmed that the binaries don't work with TF 1.15. We'll make a new release soon to fix that. Once that is done, it would be great if you could confirm that it works.

Regarding 2.0/eager mode, we are planning to make some changes to the interface of the library in order to make it more compatible with this new coding style. This will probably take at least a few weeks, though. I'll look into whether I can reproduce the problems you're reporting.

**Comment by nguerinjr** [comment]
Thanks for your attention @jonycgn. 

Just to update what I've reported:
- I've introduced the None gradient error of the entropy bottleneck. So it's not an issue. No problem to use it as long as I build the `.so`.

- The cuDNN problem remains when running your example directly. I can only run it using my code. The stack is:

``` python
2019-11-14 16:49:04.035451: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcublas.so.10.0
2019-11-14 16:49:04.225950: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcudnn.so.7
2019-11-14 16:49:04.610015: E tensorflow/stream_executor/cuda/cuda_dnn.cc:329] Could not create cudnn handle: CUDNN_STATUS_INTERNAL_ERROR
2019-11-14 16:49:04.612247: E tensorflow/stream_executor/cuda/cuda_dnn.cc:329] Could not create cudnn handle: CUDNN_STATUS_INTERNAL_ERROR
Traceback (most recent call last):
  File "/mnt/data/projects/compression/bls2017.py", line 402, in <module>
    app.run(main, flags_parser=parse_args)
  File "/mnt/data/projects/compression/venv/lib/python3.7/site-packages/absl/app.py", line 299, in run
    _run_main(main, args)
  File "/mnt/data/projects/compression/venv/lib/python3.7/site-packages/absl/app.py", line 250, in _run_main
    sys.exit(main(argv))
  File "/mnt/data/projects/compression/bls2017.py", line 390, in main
    train(args)
  File "/mnt/data/projects/compression/bls2017.py", line 160, in train
    y = analysis_transform(x)
  File "/mnt/data/projects/compression/venv/lib/python3.7/site-packages/tensorflow_core/python/keras/engine/base_layer.py", line 898, in __call__
    outputs = self.call(cast_inputs, *args, **kwargs)
  File "/mnt/data/projects/compression/bls2017.py", line 94, in call
    tensor = layer(tensor)
  File "/mnt/data/projects/compression/venv/lib/python3.7/site-packages/tensorflow_core/python/keras/engine/base_layer.py", line 898, in __call__
    outputs = self.call(cast_inputs, *args, **kwargs)
  File "/mnt/data/projects/compression/tensorflow_compression/python/layers/signal_conv.py", line 771, in call
    outputs = self._correlate_down_valid(outputs, kernel)
  File "/mnt/data/projects/compression/tensorflow_compression/python/layers/signal_conv.py", line 455, in _correlate_down_valid
    strides=self.strides_down, padding="VALID", data_format=data_format)
  File "/mnt/data/projects/compression/venv/lib/python3.7/site-packages/tensorflow_core/python/ops/nn_ops.py", line 898, in convolution
    name=name)
  File "/mnt/data/projects/compression/venv/lib/python3.7/site-packages/tensorflow_core/python/ops/nn_ops.py", line 1009, in convolution_internal
    name=name)
  File "/mnt/data/projects/compression/venv/lib/python3.7/site-packages/tensorflow_core/python/ops/gen_nn_ops.py", line 1039, in conv2d
    _six.raise_from(_core._status_to_exception(e.code, message), None)
  File "<string>", line 3, in raise_from
tensorflow.python.framework.errors_impl.UnknownError: Failed to get convolution algorithm. This is probably because cuDNN failed to initialize, so try looking to see if a warning log message was printed above. [Op:Conv2D] name: analysis_transform/layer_0/convolution/

Process finished with exit code 1
```

But that's only occurring on my PC. I've recently run your example on another computer and it was ok. So it's not a major concern. Maybe something with the libraries versions:
- Ubuntu 18.04.3 LTS, Geforce RTX 2060
- python 3
- libcudnn7 7.6.3.30-1+cuda10.0
- Nvidia driver 430.26
- TensorFlow 1.15.0

- The only problem when trying to execute eagerly with TF 1.15 are the gradients of the reparametrizations:
``` python
WARNING:tensorflow:Gradients do not exist for variables ['1_conv0/kernel_rdft:0', '1_conv0/gdn/reparam_beta:0', '1_conv0/gdn/reparam_gamma:0', '1_conv1/kernel_rdft:0', '1_conv1/gdn_1/reparam_beta:0', '1_conv1/gdn_1/reparam_gamma:0', '1_conv2/kernel_rdft:0', '1_conv2/gdn_2/reparam_beta:0', '1_conv2/gdn_2/reparam_gamma:0', '1_conv3/kernel_rdft:0', '1_latent/matrix_0:0', '1_latent/factor_0:0', '1_latent/matrix_1:0', '1_latent/factor_1:0', '1_latent/matrix_2:0', '1_latent/factor_2:0', '1_latent/matrix_3:0', '1_latent/quantiles:0', '1_deconv0/kernel_rdft:0', '1_deconv0/gdn_3/reparam_beta:0', '1_deconv0/gdn_3/reparam_gamma:0', '1_deconv1/kernel_rdft:0', '1_deconv1/gdn_4/reparam_beta:0', '1_deconv1/gdn_4/reparam_gamma:0', '1_deconv2/kernel_rdft:0', '1_deconv2/gdn_5/reparam_beta:0', '1_deconv2/gdn_5/reparam_gamma:0', '1_color/kernel_rdft:0'] when minimizing the loss.
WARNING:tensorflow:Gradients do not exist for variables ['1_conv0/kernel_rdft:0', '1_conv0/gdn/reparam_beta:0', '1_conv0/gdn/reparam_gamma:0', '1_conv1/kernel_rdft:0', '1_conv1/gdn_1/reparam_beta:0', '1_conv1/gdn_1/reparam_gamma:0', '1_conv2/kernel_rdft:0', '1_conv2/gdn_2/reparam_beta:0', '1_conv2/gdn_2/reparam_gamma:0', '1_conv3/kernel_rdft:0', '1_latent/matrix_0:0', '1_latent/factor_0:0', '1_latent/matrix_1:0', '1_latent/factor_1:0', '1_latent/matrix_2:0', '1_latent/factor_2:0', '1_latent/matrix_3:0', '1_latent/quantiles:0', '1_deconv0/kernel_rdft:0', '1_deconv0/gdn_3/reparam_beta:0', '1_deconv0/gdn_3/reparam_gamma:0', '1_deconv1/kernel_rdft:0', '1_deconv1/gdn_4/reparam_beta:0', '1_deconv1/gdn_4/reparam_gamma:0', '1_deconv2/kernel_rdft:0', '1_deconv2/gdn_5/reparam_beta:0', '1_deconv2/gdn_5/reparam_gamma:0', '1_color/kernel_rdft:0'] when minimizing the loss.
```

In this version of code, I'm using the Keras functional API and training model with the `fit` method.

**Comment by jonycgn** [comment]
We've just released new pip packages that should be compatible with TF 1.15. Can you confirm that they work for you?

Thanks for your additional information on eager mode. I will look into that in the next few weeks.

**Comment by nguerinjr** [comment]
Yes, it's working @jonycgn. Thanks for the release.

Also, I'm not having the failure I've reported earlier about loading the cuDNN algorithms. It was only happening on my computer.

I've noticed I was using Python 3.7 in my pc (which I see in your docs, it's not supported). Maybe there's something about it since now I'm using a virtualenv with Python 3.6 to install your package.

**Comment by jonycgn** [comment]
Thanks, good to hear that the cuDNN issue is resolved.

Quick question about the gradient issue: it looks like the list of missing gradients you reported contains pretty much all of the variables of the model. Are you sure that you are applying GradientTape in the correct way?

**Comment by nguerinjr** [comment]
The version of the prints is one using the fit method from the Keras model. 

I've done another with fewer layers. I'm not using Keras fit method in this version.

The main loop code is:
```python
for data in train_iter:
    with tf.GradientTape(persistent=True) as tape:
        out = model(data)
        loss = tf.math.reduce_mean(
            tf.math.squared_difference(data, out)) * (255 ** 2)

    grads = tape.gradient(loss, model.trainable_variables)
```

I was debugging here to see what's happening and you're right. Almost all variables are non-differentiable under GradientTape. Only biases are differentiable.

![trainable_variables](https://user-images.githubusercontent.com/35977339/69828129-54798380-11f9-11ea-9fde-29393bdf5099.png)

![grads_of_trainable_variables](https://user-images.githubusercontent.com/35977339/69828133-57747400-11f9-11ea-9042-5ec23c16a41f.png)

Even though I'm not using Keras fit method, I'm prototyping with the Keras model. So maybe there's something about it. I don't know yet.

**Comment by jonycgn** [comment]
Ok, I think I've reproduced the gradient problem with eager mode. The problem seems to be the reparameterization. Can you try to pass `kernel_reparametizer=None` to `SignalConv2D` and see if that helps?

**Comment by jonycgn** [comment]
Commit 19eb16cbde0939c9953ed2035c6cb71eb57981de should fix the gradients for Keras layers (SignalConv and GDN).

**Comment by nguerinjr** [comment]
Hello @jonycgn. I'll test it here and report it to you in the next days. I've been very busy. That's why I didn't answer before.

**Comment by jonycgn** [comment]
No problem.

I've also just added a re-implementation of the entropy models specifically for eager mode (see commit b996d5c277a3957b7be0c578e5d1bc4f12afe330). There simply isn't a way to port the old entropy models to eager mode, and the Keras layer interface was quite difficult to support to begin with.

The new code is still work in progress – if you try it out, it would be great if you could post any feedback here!

**Comment by jonycgn** [comment]
As of today, we have a beta release out that supports TF2 and eager mode.

**Comment by moha23** [comment]
`pip install tensorflow-compression==2.0b1` throws error
```
ERROR: Could not find a version that satisfies the requirement tensorflow-compression==2.0b1
ERROR: No matching distribution found for tensorflow-compression==2.0b1
```
on Ubuntu20.04,Tf2.3.1

**Comment by jonycgn** [comment]
@moha23
Please run the command `pip search tensorflow-compression`. Does it print the following lines (among other items)?
```
tensorflow-compression (1.3)                       - Data compression in TensorFlow
  INSTALLED: 2.0b1
  LATEST:    1.3
```
Also, please paste the output of the command `uname -a` here.

**Comment by moha23** [comment]
it shows the following without `installed` and `latest` info:

```
tensorflow-compression (1.3)                       - Data compression in
                                                     TensorFlow
```
System info:

`Linux mohana-C7Z170-SQ 5.4.0-54-generic #60-Ubuntu SMP Fri Nov 6 10:37:59 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux`

**Comment by jonycgn** [comment]
What's your Python version? Note that only 3.5--3.8 is supported.

**Comment by nguerinjr** [comment]
Hello @jonycgn.

I've performed a simple environment test. I could install the beta version in python3.7. But not in python 3.8

![Captura de tela 2020-10-30 143616](https://user-images.githubusercontent.com/35977339/101369927-9ebf2700-3887-11eb-9c38-d4c58957b1ac.png)

In 3.7 it works perfectly. At least analyzing only the pip outputs.

**Comment by jonycgn** [comment]
Thanks @nguerinjr! What error message are you getting in 3.8?

**Comment by moha23** [comment]
Same, the error is with python3.8. Installs fine with python3.6.

**Comment by jonycgn** [comment]
Please download this file: https://files.pythonhosted.org/packages/e9/8e/8dd972392acb5f8b9b40b39761922a12b1c6ad554de4bb9887e3e8f736d3/tensorflow_compression-2.0b1-cp38-cp38m-manylinux2010_x86_64.whl

And try to install it directly. Does that work?

**Comment by nguerinjr** [comment]
> Thanks @nguerinjr! What error message are you getting in 3.8?

@jonycgn, there's no problem in 3.6 and 3.7 versions. I've done the tests again just to confirm it. In 3.8 the error is the following:

![Captura de tela 2020-10-30 143616](https://user-images.githubusercontent.com/35977339/101399272-01c3b480-38ae-11eb-9f15-e56d3532d404.png)

**Comment by jonycgn** [comment]
Hi, can you please try to download the file I linked above, and then do `pip install downloaded_file` manually? Thanks.

**Comment by nguerinjr** [comment]
This is the output I had. There's something strange. It's saying it's not a supported wheel on this platform.

![Captura de tela 2020-10-30 143616](https://user-images.githubusercontent.com/35977339/101402556-a811b900-38b2-11eb-8451-96da9c99d5ce.png)

**Comment by jonycgn** [comment]
That is odd. Maybe something went wrong in the way we were building the package…

**Comment by nguerinjr** [comment]
I'll perform some tests of the beta version of the package using python 3.7. I have a bunch of codes that use your framework. I'll report if I find any bugs.

When this problem of the 3.8 version is fixed I can migrate to this version of python.

**Comment by jonycgn** [comment]
Sounds good. As a start, I'd recommend running the unit tests: `python -m tensorflow_compression.python.all_test`. Thanks!

I will try to find out what happened with 3.8.

**Comment by nguerinjr** [comment]
I've performed these unit tests on two machines with two RTX 2080 Ti. Performed both on CPU and GPU. The unit tests were all ok. Now I'll start implementing a 2.0 version of some codes using the beta framework.

I'm available to help you with tests, or perform new tests in the machines I use if you want so.

**Comment by nguerinjr** [comment]
@jonycgn, I have the following error on entropy_models.py

![Captura de tela 2020-12-07 190142](https://user-images.githubusercontent.com/35977339/101521690-eff11880-3964-11eb-85d6-a922260ee1e8.png)

The entropy models are now working in TF 2.0 eager mode. I think it's just a test that should not be here anymore, isn't it?

Do you prefer that I relate the bugs here in this thread? Or open new threads to handle each 2.0 beta bug I find?

**Comment by nguerinjr** [comment]
Considering a code with this idea, when I wrap the layers with the Keras functional API

``` python

  # Instantiate Keras model.
  keras_input = tf.keras.Input(batch_size=args.batch_size, shape=[args.block_size] * 3 + [1], dtype=tf.float32)
  analysis_out = AnalysisTransform(args.num_filters)(keras_input)
  entropy_y, entropy_likelihood = tfc.EntropyBottleneck()(analysis_out, training=True)
  synthesis_out = SynthesisTransform(args.num_filters)(entropy_y)
  k_model = tf.keras.Model(inputs=[keras_input], outputs=[synthesis_out, entropy_likelihood])
  avulse_bottleneck = tfc.EntropyBottleneck()

  # Instantiate optimizers
  main_optimizer = tf.keras.optimizers.Adam(learning_rate=1e-4)
  aux_optimizer = tf.keras.optimizers.Adam(learning_rate=1e-3)

  # Main loop
  for step, cloud in enumerate(cloud_input):
    with tf.GradientTape(persistent=True) as g_tape:
      rec_cloud, likelihoods = k_model(cloud)  # Model running
      res_avulse_bottleneck = avulse_bottleneck(cloud, training=True)
      wbce_loss = wBCE_loss(args.pos_weight, cloud, rec_cloud)
      bpv_loss = tf.reduce_sum(tf.math.log(likelihoods)) / (-np.log(2) * tf.reduce_sum(cloud))
      train_loss = wbce_loss
      if args.lmbda > 0:  # Variational
        train_loss *= args.lmbda
        train_loss += bpv_loss

      # I can access the losses from avulse_bottleneck
      aux_loss = avulse_bottleneck.losses[0]
      # I can't acess the losses when wrapped in keras model (it's empty)
      k_model.layers[2].losses[0]  
```

If I instantiate the bottleneck without wrapping it in a Keras model, I can access it's attribute losses

But, when accessing it using a corresponding layer of the Keras model, I can't. Is the framework intended to allow to be used like this, wrapping with a Keras model, or is it intended to work only instantiating the components manually and calling them manually?

If this flexibility is desirable, maybe there could be a calc_loss function to calc it at that moment, for example. It's an idea.

Another thing is that the way calling bottleneck.updates[0] doesn't seem to work anymore. The updates field, in both cases, is empty. In this case, I think a function update would be a good idea. When someone calls this function it could update the CDF of the bottleneck.

What do you think? I was trying to make a version more flexible here that could work in these situations.

**Comment by nguerinjr** [comment]
Another point @jonycgn. Even using the code in the default way, when I can get the losses component of the bottleneck:

``` python
  analysis = AnalysisTransform(args.num_filters)
  bottleneck = tfc.EntropyBottleneck()
  synthesis = SynthesisTransform(args.num_filters)
  # Main loop
  for step, cloud in enumerate(cloud_input):
    with tf.GradientTape(persistent=True) as g_tape:
      y = analysis(cloud)
      y_tilde, likelihoods = bottleneck(y, training=True)
      x_tilde = synthesis(y_tilde)

      wbce_loss = wBCE_loss(args.pos_weight, cloud, x_tilde)
      bpv_loss = tf.reduce_sum(tf.math.log(likelihoods)) / (-np.log(2) * tf.reduce_sum(cloud))
      train_loss = args.lmbda * wbce_loss + bpv_loss + bottleneck.losses[0]

    grads = g_tape.gradient(train_loss, analysis.trainable_weights + bottleneck.trainable_weights + synthesis.trainable_weights)
    names = list(map(lambda w: w.name, analysis.trainable_weights + bottleneck.trainable_weights + synthesis.trainable_weights))
    zip(*[grads, names])
```

There are None gradients, as I show below. Mainly related to the functions that compose the CDF calculation.
![Captura de tela 2020-12-07 190142](https://user-images.githubusercontent.com/35977339/101552374-83d8d980-3991-11eb-9eb1-3194a712f6fc.png)

I don't know if there's a type of error in the idea of the codes I've pasted here.

But I'm analyzing the framework code to see if I can make some modifications to work around it. If that's the case I can show you  later the changes I'm thinking of.

**Comment by nguerinjr** [comment]
An addendum: I think to continue using the add_loss in the build will not work. It works firstly because it's calculated in the build once it's called in the instantiation. 

But in the data loop, when I access the bottleneck.losses[0] again, it's empty. 

So I think a good way would be to create a function "calc_loss". I've changed the code using this approach and now I can get gradients for the quantiles at least. And it works whenever you're wrapping or not with Keras model.

I'm looking for the reason of the other None gradients.

**Comment by nguerinjr** [comment]
Hello @jonycgn. Sorry for the number of messages. I've been reporting everything I've found.

I have analyzed each of the issues and I have understood what's the deal with TF 2.0. I have implemented modifications and now I have a working version of the framework in all the cases.

If you want, I can make a pull request so you can see what changes I've made.

**Comment by jonycgn** [comment]
Hi nguerinjr,

sorry for the delayed response. The entropy models in [entropy_models.py](https://github.com/tensorflow/compression/blob/master/tensorflow_compression/python/layers/entropy_models.py) are not intended to be used in TF 2.0. We have new implementations in [python/entropy_models](https://github.com/tensorflow/compression/tree/master/tensorflow_compression/python/entropy_models). Please take a look at them! There is one example model that uses them already [here](https://github.com/tensorflow/compression/blob/master/models/ms2020.py). We will convert the other models in time.

**Comment by jonycgn** [comment]
Everyone, I just addressed a problem in the beta release pip files. This should fix the "no matching distribution found" error. To summarize:

With this, you should get the stable version:
pip install tensorflow==1.15 tensorflow-compression==1.3

This should give you the beta version:
pip install tensorflow==2.3 tensorflow-compression==2.0b1

The latter should work for Python 3.6 to 3.8, for both MacOS and Linux. For the former, it will depend on the Python version.

**Comment by nguerinjr** [comment]
> Hi nguerinjr,
> 
> sorry for the delayed response. The entropy models in [entropy_models.py](https://github.com/tensorflow/compression/blob/master/tensorflow_compression/python/layers/entropy_models.py) are not intended to be used in TF 2.0. We have new implementations in [python/entropy_models](https://github.com/tensorflow/compression/tree/master/tensorflow_compression/python/entropy_models). Please take a look at them! There is one example model that uses them already [here](https://github.com/tensorflow/compression/blob/master/models/ms2020.py). We will convert the other models in time.

Fine. I'll have a look at the example and at the new implementations. What I've done earlier was to adapt the entropy_models.py. I didn't pay attention to other files. Thanks, @jonycgn.

**Comment by jonycgn** [comment]
Note, I just bumped the beta release to v2.0b2, which should now work with TF 2.4.

I'm going to close this issue, since it was initially opened about running entropy models in TF2. We now have the new implementations, which should fix this.

If there are any other issues with either installing pip packages, or using the new entropy model implementations, let's create a new issue for these as needed.

Thanks everyone for your inputs!
